apps/files/lib/Command/ListFiles.php

```diff
@@ -0,0 +1,311 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program. If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+namespace OCA\Files\Command;
+
+use OC\Core\Command\Base;
+use OC\Core\Command\InterruptedException;
+use OC\FilesMetadata\FilesMetadataManager;
+use OC\ForbiddenException;
+use OCP\EventDispatcher\IEventDispatcher;
+use OCP\Files\File;
+use OCP\Files\FileInfo;
+use OCP\Files\Folder;
+use OCP\Files\IRootFolder;
+use OCP\Files\NotFoundException;
+use OCP\IUserManager;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\Console\Helper\Table;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class ListFiles extends Base {
+	protected array $fileInfo = [];
+	protected array $dirInfo = [];
+	public function __construct(
+		private IUserManager $userManager,
+		private IRootFolder $rootFolder,
+		private FilesMetadataManager $filesMetadataManager,
+		private IEventDispatcher $eventDispatcher,
+		private LoggerInterface $logger
+	) {
+		parent::__construct();
+	}
+
+	protected function configure(): void {
+		parent::configure();
+
+		$this->setName("files:list")
+			->setDescription("List filesystem in the path mentioned in path argument")
+			->addArgument(
+				"path",
+				InputArgument::REQUIRED,
+				'List all the files and folder mentioned in this path, eg occ files:list path="/alice/files/Music", the path being a required argument to determine the user'
+			)
+			->addOption("type", "", InputArgument::OPTIONAL, "Filter by type like application, image, video etc")
+			->addOption(
+				"minSize",
+				'0',
+				InputArgument::OPTIONAL,
+				"Filter by min size"
+			)
+			->addOption(
+				"maxSize",
+				'0',
+				InputArgument::OPTIONAL,
+				"Filter by max size"
+			)
+			->addOption(
+				"sort",
+				"name",
+				InputArgument::OPTIONAL,
+				"Sort by name, path, size, owner, type, perm, created-at"
+			)
+			->addOption("order", "ASC", InputArgument::OPTIONAL, "Order is either ASC or DESC");
+	}
+
+	private function getNodeInfo(File|Folder $node): array {
+		$nodeInfo = [
+			"name" => $node->getName(),
+			"size" => $node->getSize() . " bytes",
+			"perm" => $node->getPermissions(),
+			"owner" => $node->getOwner()->getDisplayName(),
+			"created-at" => $node->getCreationTime(),
+			"type" => $node->getMimePart(),
+		];
+		if($node->getMimetype() == FileInfo::MIMETYPE_FOLDER) {
+			$nodeInfo['type'] = 'directory';
+		}
+		
+		return $nodeInfo;
+	}
+
+	protected function listFiles(
+		string $user,
+		string $path,
+		OutputInterface $output,
+		?string $type = "",
+		?int $minSize = 0,
+		?int $maxSize = 0
+	): void {
+		try {
+			/** @ var OC\Files\Node\Folder $userFolder **/
```

Comment:
Psalm is complaining because `$this->rootFolder->get` return type is `Node`, so we need to tell psalm that we know what we are doing, and the `$userFolder` is a `Folder`. I think the following should work:

```suggestion
			/** @var OCP\Files\Folder $userFolder **/
```

That, or add the following check:

```php
if (!($userFolder instanceof Folder)) {
    // throw or return.
}
```